Adds glossary sub-page for nodes (#411)
BitcoinDesign/Guide: guide/glossary/node.md

   alt-text = "Network graph of Bitcoin node connections"
   width = 1600
   height = 500
   layout = "full-width"
%}

# Node

A node refers to a Bitcoin client that participates in the Bitcoin network. Any Bitcoin application that wants to show and send transactions needs a way to connect to a node and manage block data and there are many ways to make this work.

There are multiple aspects that differentiate nodes:
- Is it controlled by you or someone else? How much trust is involved?
- Does it verify the data it receives?
- Does it store all transaction data or only data relevant to you?
- Security considerations, such as the risk of receiving manipulated data
- Privacy considerations, such as tracing your transaction behavior

## Common node setups

From complete control to fully trusting a single third party, there are a variety of node setups. Below are examples of common ones.

#### Local node

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/local-node.png"
   retina = "/assets/images/guide/glossary/node/local-node@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

> A very simple setup is to use an application like Bitcoin Core that combindes node and wallet functionality.

Reviewer: ```suggestion
A very simple setup is to use an application like Bitcoin Core that combines node and wallet functionality.
```

</div>

#### Hardware node

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/hardware-node.png"
   retina = "/assets/images/guide/glossary/node/hardware-node@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

> Separating node and wallet functionality, many users rely on a dedicated computer like a Rasperry Pi running software like Umbrel, MyNode, and Raspiblitz. Wallet applications can connect to the node which allows them to be more lightweight.

Reviewer: ```suggestion
Separating node and wallet functionality, many users rely on a dedicated computer like a Raspberry Pi running software like Umbrel, MyNode, and Raspiblitz. Wallet applications can connect to the node which allows them to be more lightweight.
```

</div>

#### Self-hosted node

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/own-hosted-node.png"
   retina = "/assets/images/guide/glossary/node/own-hosted-node@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

> While it possible to run a node on a rented server, this partially places the node in the control of a third party and introduces privacy risks.

Reviewer: So a self hosted node is one which someone else hosts?

Reviewer 2: Should change this to 'Hosted node.' It's your own node but on someone else's server. BTCPay using Luna is an example

Author: I had added "self", as you configure and host it on someone else's generic servers. To me, a "Hosted node" would be something like https://www.quicknode.com, where I just press a button and they spin up a node for me.

Author: It was more about the relationship for me here. You control the software, but the hardware belongs to someone else.

</div>

#### Third-party node

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/third-party-node.png"
   retina = "/assets/images/guide/glossary/node/third-party-node@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

> A convenient setup is to connect to a node controlled by a third-party, which is only recommended if there is a good trust relationship.

Reviewer: What is the extent of control mean? Who owns this third party node? Can I own it, or partially own it?

Author: It's meant as the node being out of your hands altogether - you have no control over it and may not even know who owns it.

Reviewer: Ok and the reason you would be using such node is so that you can broadcast your transactions, and retrieve block data.

Otherwise there are no contributions to the network as you are not validating transactions or mining. You are trusting that the node operator gives you the correct data. As this can be a point of centralisation if too many people rely on such node operators it is usually discouraged.

Author: Exactly. It's maybe fine if a family member is controlling the node, or if you're in some other formal trust relationship like the community-managed setup used by Galoy. Good for starters or when the economic cost of running a node is too high and there's good trust relationship. It has its place, but should be considered carefully.

Reviewer: > This setup is good for starters or when the economic cost of running a node is too high and there's good trust relationship. It is primarily used to broadcast transactions, and retrieve block data otherwise there are no contributions to the network as you are not validating transactions or mining. You are trusting that the node operator gives you the correct data, you have no control over the node and may not even know who operates it. As this can be a point of centralisation if too many people rely on such node operators it is usually discouraged.

Author: Great. Wanna post it as a suggestion and I'll merge?

Reviewer: Well that could probably could use some more word smithing. If you think there's anything useful in there just copy and adapt.

Author: Just rewrote the text with the points you made.

</div>

> #### Multiple third-party nodes

Reviewer: Multiple third-party nodes, and Third-party API talk primary about connections not node setups.

Author: Setup is meant in the sense of "arrangement" - the relationship between user, wallet and node. I found it tricky to find the right umbrella term here, let me know if you have a better idea.

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/third-party-node.png"
   retina = "/assets/images/guide/glossary/node/third-party-node@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

> Connecting to a single node controlled by a third-party is a convenient setup for starters, or when the economic cost of running a node is too high. This should only be used if there is a good trust relationship or potential loss is minimal. Too many user relying on third-party nodes can lead to centralization, so it is generally discouraged.

Reviewer: looks good. only thing I'd ask is what does a potential loss look like for on-chain third party node? besides leaking some identifying information?

Author: Good question. They could block your transactions and feed you with fake data to mess with you. Not sure if that power could then be used to get to your funds...

Reviewer: So an inconvenience. Loss of time not loss of funds.

</div>

#### Third-party API

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/third-party-api.png"
   retina = "/assets/images/guide/glossary/node/third-party-api@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

> Third-party APIs should be avoided. These are custom connection mechanisms built on top of block data, and there is no way to validate the authenticity of the data.

Reviewer: This one can also be seen to contradict the previous heading. At least a reorder could help.

> Connecting to multiple untrusted third-parties can be a good idea if compact block filters are used

> Third-party APIs should be avoided. These are custom connection mechanisms built on top of block data, and there is no way to validate the authenticity of the data.

Author: The difference is API vs. node. A node sends you actual block data while an API can modify and reorganize data and send you something different. Does that address your concern? Or did I not understand correctly?

Reviewer: Also these two headings specifically touch on types of node connections vs setup.

Author: I flipped the order slightly so the options we don't recommend are at the bottom. Let me know if that addresses your feedback.

</div>

## Block data management

> Based on use case and context, some nodes are very thorough and store and verify the complete block chain, while others are optimized for efficiency and may accept slight security and privacy tradeoffs.

Reviewer: ```suggestion
Based on use case and context, some nodes are very thorough and store and verify the complete blockchain, while others are optimized for efficiency and may accept slight security and privacy tradeoffs.
```

#### Full nodes

Full nodes are Bitcoin clients that fully verify and enforce the rules of the Bitcoin network. An example of a rule is that there will never be more than 21 million bitcoin.

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/archival-node.png"
   retina = "/assets/images/guide/glossary/node/archival-node@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

> **Archival full nodes** store the entire Bitcoin block chain, and send historical data to other nodes.

Reviewer: ```suggestion
**Archival full nodes** store the entire Bitcoin blockchain, and send historical data to other nodes.
```

   width = 100
   height = 100
   layout = "float-left"
%}

**Pruned full nodes** use much lower storage capacity as they only store a portion of recent blocks.

</div>

#### Light nodes

Light nodes are Bitcoin clients that do not verify or enforce the rules of the Bitcoin network.

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/spv-node.png"
   retina = "/assets/images/guide/glossary/node/spv-node@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

**Simplified payment verification (SPV) light nodes** query and download block headers from full nodes making it possible for users to verify transactions without running a full node.

For more information:

   - [Bitcoin Wiki SPV](https://en.bitcoinwiki.org/wiki/Simplified_Payment_Verification)
   - [Bitcoin whitepaper - section 8](https://bitcoin.org/bitcoin.pdf)

</div>

<div class="center" markdown="1">

{% include image.html
   image = "/assets/images/guide/glossary/node/bloom-filter-node.png"
   retina = "/assets/images/guide/glossary/node/bloom-filter-node@2x.png"
   alt-text = ""
   width = 100
   height = 100
   layout = "float-left"
%}

> **Neutrino light nodes** uses the Neutrino protocol. The Neutrino protocol uses compact block filters to query and download condensed representations of blocks that contain transaction data relevant to the user.

Reviewer: ```suggestion
**Neutrino light nodes** use the Neutrino protocol. The Neutrino protocol uses compact block filters to query and download condensed representations of blocks that contain transaction data relevant to the user.
```
